src/main/kotlin/org/rust/ide/annotator/RsFileErrorAnnotator.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.annotator
+
+import com.intellij.codeInsight.daemon.impl.HighlightRangeExtension
+import com.intellij.ide.annotator.AnnotatorBase
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.cargo.project.model.cargoProjects
+import org.rust.cargo.project.model.impl.allTargets
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.RsInnerAttr
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.descendantsOfType
+import org.rust.lang.utils.RsDiagnostic
+import org.rust.lang.utils.addToHolder
+
+class RsFileErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
+
+    override fun isForceHighlightParents(file: PsiFile): Boolean = file is RsFile
+
+    override fun annotateInternal(element: PsiElement, holder: AnnotationHolder) {
+        val visitor = object : RsVisitor() {
+            override fun visitFile(file: PsiFile) { checkRsFile(holder, file as? RsFile ?: return) }
+        }
+
+        element.accept(visitor)
+    }
+
+    private fun checkRsFile(holder: AnnotationHolder, file: PsiFile) {
+        val binCargoProject = file.project.cargoProjects.allTargets.lastOrNull { target -> target.isBin && file.virtualFile == target.crateRoot } ?: return
+
+        if (file.descendantsOfType<RsInnerAttr>().count { rsInnerAttr -> "no_main" == rsInnerAttr.metaItem.text } == 1) return
+        if (file.descendantsOfType<RsFunction>().lastOrNull { fn -> "main" == fn.name } != null) return
```

Comment:
`descendantsOfType` scans all chindren in AST (not only direct). So if you have any function with `main` name in file, it finds it. But it but that is not what we want